src/batch_write.rs

```diff
@@ -59,27 +59,27 @@ impl DbInner {
             let mut guard = self.state.write();
 
             guard.update_clock_tick(now)?;
+            let seq = guard.increment_seq();
```

Comment:
should this be incremented with every op and not just once for the whole batch? 🤔

I think it feels semantically correct as long as batch is atomic, but I still didn't expect it.

If it's correct (which I suspect it is), should we have a test that ensures this contract won't regress?

Reply:
🤔 should we disallow multiple merges in the same batch? (in case MergeOperator returns error) 

This would mean pushing conflict resolution errors within the batch on the user, which feels legit. For non-batched write we delegate error handling to database.

https://github.com/slatedb/slatedb/blob/main/rfcs/0006-merge-operator.md#error-handling

Reply:
> should this be incremented with every op and not just once for the whole batch? 🤔

no it's by design, implementing transaction requires the batch have the same seqnum. this `seqnum` is sometimes called as `commit timestamp` or `commit version` in other systems, it's a single timestamp/version of a single transaction commit.

Reply:
> 🤔 should we disallow multiple merges in the same batch? (in case MergeOperator returns error)
> 
> This would mean pushing conflict resolution errors within the batch on the user, which feels legit. For non-batched write we delegate error handling to database.
> 
> https://github.com/slatedb/slatedb/blob/main/rfcs/0006-merge-operator.md#error-handling

oh sorry i haven't learned the details about the design of merge before, let me learn about it. thank you for the info!

Reply:
Hmm, there are also cases such as multiple puts with different expiration timestamps (later write expiring first).

It feels like having a single seq number per batch would be rather limiting, because there are cases where you can end up with multiple versions of the same key in the same batch 🤔

Reply:
but this behaviour is shared in all the LSM engines 😲

Reply:
I think we simply need to limit write batch api and enforce single merge entry per-key. That feels to be a fair limitation.

> there are also cases such as multiple puts with different expiration timestamps (later write expiring first).

this case turns out to be invalid; later one would turn into tombstone

Reply:
> I think we simply need to limit write batch api and enforce single merge entry per-key. 

i think it's allowed that a key can have multiple write operations in a write batch. if take merge operations into consideration, I guess it may be necessary to treat the operations within the write batch as an ordered set. however, I haven't studied in detail how the write batch handles multiple writes to the same key in other systems yet.

how about let's put this constraint in another PR? i'd like to make some further investigations about these details in the WriteBatch, as it's important to get the WriteBatch implementation right for making transactions. however, I hope this PR focuses on adding the seqnum, so we can continue the work on both snapshot and merge. wdyt?

Reply:
Sounds good 👍